### PR TITLE
Add JsProxy.to_py method

### PR DESCRIPTION
**Pull request by hoodmane**
Currently converts `Map` to `dict`, `Set` to `set` and array-like objects to `list`. If the object is a base `Object` (not a subclass of `Object`, I check that `Object.constructor.name === "object"` and that it doesn't have a custom `Symbol.toStringTag`) then I convert it to a `dict` as well. Otherwise I leave it alone.

Takes a depth parameter which works in the same way as `PyProxy.toJs`.

### CODE REVIEW

**Comment by hoodmane** [comment]
Here's one limitation of the current implementation:
```javascript
z = new Map();
z.set([1,1], 2);
pyodide.runPython(`
    from js import z
    z.to_py()
`)
```
Raises `TypeError: unhashable type: 'list'`.

**Comment by rth** [comment]
> Here's one limitation of the current implementation:
> TypeError: unhashable type

I think that's an expected error from the Python perspective on unhashable keys, so better to raise it than try to find some non obvious workaround (such as converting keys to str).

File: src/tests/test_typeconversions.py [hunk 1]
```diff
@@ -441,6 +441,69 @@ def test_python2js_with_depth(selenium):
...
+            from js import a
+            b = a.to_py()
+            repr(type(b))
+        `);
+        return result;
+        """
+    )
+    assert result == "<class 'JsProxy'>"
+
+
```
Comment (phorward) [inline src/tests/test_typeconversions.py:506]: Can you add your failing conversion tests from [here](https://github.com/iodide-project/pyodide/pull/1244#issuecomment-778386662), so it shows that we already know about these errors and don't do any "falsification" on JavaScript objects to fit into the Python world?

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @hoodmane,
thanks for your work! I'm mostly well with this pull request, but what do you think about adding the failing tests as well to the test suite? See my last comment above.

**Comment by hoodmane** [comment]
@phorward Thanks for the review!

File: src/core/js2python.c [hunk 2]
```diff
@@ -124,39 +133,174 @@ EM_JS_REF(PyObject*, __js2python, (JsRef id), {
...
+      let entryid = Module.hiwire.new_value(obj[i]);
+      let item = Module.__js2python_convert(entryid, map, depth);
+      Module.hiwire.decref(entryid);
+      // clang-format off
+      if (item === 0) {
+        // clang-format on
+        _Py_DecRef(list);
+        return 0;
+      }
+      // PyList_SetItem steals a reference to item no matter what
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/core/js2python.c:178]: I think it's helpful to have this comment here, without it at some point I'm going to think "Why didn't I decref `item` isn't that a reference leak?" and have to look up `PyList_SetItem` in the docs to find out that it steals a reference to its argument. (I actually had to look at the source of `PyList_SetItem` to check that it always `DECREF`s the item, even if an early error occurs.)

**Review by phorward — APPROVED** [review]
OK, I'm fine with this, thanks for updating the tests!

**Comment by rth** [comment]
Thanks! I have not reviewed the JS code in detail but the tests look good. Merging with a +1.

**Pull request merged**